lib/engine/chelsea_bridge.ex

```diff
@@ -22,36 +26,84 @@ defmodule Engine.ChelseaBridge do
-         %{"bridge" => %{"bridgeStatusId" => %{"status" => status}}} <- data do
-      estimate =
-        case data do
-          %{"lift_estimate" => %{"estimate_time" => estimate}} ->
-            String.replace(estimate, ~r/\.\d+$/, "")
-            |> Timex.parse!("{YYYY}-{M}-{D} {h24}:{m}:{s}")
-            |> DateTime.from_naive!("America/New_York")
-
-          _ ->
-            nil
-        end
+           http_client.get("#{@base_api_url}#{@api_status_endpoint}", [
+             {"Authorization", "Bearer #{get_cached_or_new_token(now)}"}
+           ]),
+         {:ok, data} <- Jason.decode(body) do
+      # TODO: Estimate maybe should be nil in cases when estimatedDurationInMinutes is 0
+      :ets.insert(
+        :bridge_status,
+        {:value,
+         %{
+           raised?: Map.get(data, "liftInProgress"),
+           estimate: DateTime.add(now, Map.get(data, "estimatedDurationInMinutes"), :minute)
+         }}
+      )
 
-      :ets.insert(:bridge_status, {:value, %{raised?: status == "Raised", estimate: estimate}})
+      {:noreply, state}
     else
       err ->
         Logger.error("Error getting bridge status: #{inspect(err)}")
     end
-
-    {:noreply, state}
   end
 
   def handle_info(msg, state) do
     Logger.warning("Engine.ChelseaBridge unknown_message: #{inspect(msg)}")
     {:noreply, state}
   end
+
+  defp get_cached_or_new_token(now) do
+    case :ets.lookup(:bridge_status, :auth) do
+      [auth: %{access_token: access_token, expiration: expiration}] ->
+        # If token found, only return if still before expiration date
+        if DateTime.compare(expiration, now) == :gt do
+          access_token
+        else
+          update_api_token(now)
+        end
+
+      _ ->
+        update_api_token(now)
+    end
+  end
+
+  def update_api_token(now) do
+    # Calls a token granting endpoint with a username and password from our environment variables.
+    # This token grants Bridge Data API access, and must be refreshed at least every month
+    username = Application.get_env(:realtime_signs, :chelsea_bridge_username)
+    password = Application.get_env(:realtime_signs, :chelsea_bridge_password)
+    http_poster = Application.get_env(:realtime_signs, :http_poster_mod)
+
+    # Encode as application/x-www-form-urlencoded
+    body =
+      URI.encode_query(%{
+        "grant_type" => "password",
+        "username" => username,
+        "password" => password
+      })
+
+    with {:ok, %{status_code: 200, body: body}} <-
+           http_poster.post("#{@base_api_url}#{@api_token_endpoint}", body, [
+             {"Content-Type", "application/x-www-form-urlencoded"}
+           ]),
+         {:ok, data} <- Jason.decode(body) do
+      expiration =
+        now
+        |> DateTime.add(Map.get(data, "expires_in"), :second)
+
+      :ets.insert(
+        :bridge_status,
+        {:auth, %{access_token: Map.get(data, "access_token"), expiration: expiration}}
+      )
```

Comment:
Note: We use ETS in these engines so that the sign processes can read the data in parallel without creating a bottleneck on the engine process. Since the cached token is only used internally, it would be preferable to store it in state.

Reply:
Did modify this token and its expiration to be saved in state!